### PR TITLE
Add support for sparse-checkout

### LINKED ISSUE #54: Add sparse checkout feature
Hi there,

Would be nice if we could sparse checkout a specified list of folders/files.

### PR DESCRIPTION
**Pull request by SpencerMalone**
Wanted to finally tackle https://github.com/kubernetes/git-sync/issues/54, I sidestepped the problem of how to handle the volume of flags that might be required by instead specifying a sparsecheckout file.

The workflow as I've had has been...
- Do a local sparse checkout, add the files you want ignored (or included on if you did a cone pattern https://git-scm.com/docs/git-sparse-checkout#_cone_pattern_set)
- Grab your .git/info/sparecheckout file, and reserve it for later use with this new flag

It's not quite as easy as specifying it all from a CLI, but I think it's a reasonable first pass.

Here are some logs of it being run on https://github.com/SpencerMalone/logstash-output-prometheus:

```
test-repo % cat sparseconfig 
!/*
!/*/
README.md
test-repo % docker run --rm -d \
    -v $(pwd)/git-data:/tmp/git \
    -v $(pwd):/test \
    docker.io/registry/git-sync:tag__linux_amd64  \
        --repo=https://github.com/SpencerMalone/logstash-output-prometheus.git \
        --branch=master \
        --sparse-checkout=/test/sparseconfig
41494548dd64caf0ff8f7b75e4d3a86014cfaefc40ff31b14ba19accf99aa82f
test-repo % ls git-data/db86200b1ab158ce9ad403d06de2301b15333601 
README.md
``` 

As you can see, I ignored everything but the `README.md`, and sure enough only got that file in my final checkout.

I also created https://github.com/SpencerMalone/testrepo and did a similar test, creating commits for my included and excluded files to ensure that it didn't create any unexpected files when updating. Everything seeeeemed to work correctly, but so far I have personally not tested this in a production environment.

Fixes #54

### CODE REVIEW

**Comment by SpencerMalone** [comment]
I think I verified now? I had filled it out before, but used a different email from my committer email as it said `The e-mail address is not made public and will only be used if you wish to receive a new password or wish to receive certain news or notifications by e-mail.`, sadness.

**Comment by SpencerMalone** [comment]
I signed it

**Comment by thockin** [comment]
I have never used this aspect of git before, so I don't really know how the UX should feel.  This might be reasonable or not - I wonder how we can solicit any more feedback on it?

**Review by thockin — COMMENTED** [review]
Code looks reasonable, if we are confident that a whole-file as input is acceptble.

Can you please add a test_e2e.sh case for this?

**Comment by SpencerMalone** [comment]
I'll get the test and minor changes in today, hopefully.

I agree re: "is this the best UX". I am fairly confident that the whole-file is an input that people will want if they have complicated rules, but I think there is likely an easier entrypoint for simple cases / rulesets. When I looked at it, I got very hung up on how to handle environment variables and flags as entry points if you needed to have multiple rules, so I decided that it would be worth starting with the whole file, and trying to come up with a friendlier CLI only approach in a later iteration of the feature (similar to how git themselves seem to have rolled this out).

**Comment by SpencerMalone** [comment]
Updated with the requested changes, included an e2e test run, took some finagling to get the e2e tests happy on a mac, but it wasn't too bad and it passed in my testing.

File: README.md [hunk 1]
```diff
@@ -94,6 +94,7 @@ docker run -d \
...
 | GIT_SYNC_MAX_SYNC_FAILURES      | `--max-sync-failures`      | the number of consecutive failures allowed before aborting (the first sync must succeed, -1 will retry forever after the initial sync)                                                                                                        | 0                             |
 | GIT_SYNC_PERMISSIONS            | `--change-permissions`     | the file permissions to apply to the checked-out files (0 will not change permissions at all)                                                                                                                                                 | 0                             |
+| GIT_SPARSE_CHECKOUT            | `--sparse-checkout`         | the location of an optional [sparse checkout](https://git-scm.com/docs/git-sparse-checkout#_sparse_checkout) file, same syntax as a .gitignore file.                                                                    | ""                             |
```
Comment (thockin) [inline README.md:97]: This is out of sync with code

File: test_e2e.sh [hunk 2]
```diff
@@ -1198,6 +1198,38 @@ assert_file_eq "$ROOT"/link/file "$TESTCASE"
 # Wrap up
 pass
 
+##############################################
+# Test spare-checkout files
```
Comment (thockin) [inline test_e2e.sh:1202]: typo: sparse

**Review by thockin — COMMENTED** [review]
Thanks!  super tiny things left.

**Comment by SpencerMalone** [comment]
Ah, TY for the final catches!

File: README.md [hunk 3]
```diff
@@ -94,6 +94,7 @@ docker run -d \
...
 | GIT_SYNC_MAX_SYNC_FAILURES      | `--max-sync-failures`      | the number of consecutive failures allowed before aborting (the first sync must succeed, -1 will retry forever after the initial sync)                                                                                                        | 0                             |
 | GIT_SYNC_PERMISSIONS            | `--change-permissions`     | the file permissions to apply to the checked-out files (0 will not change permissions at all)                                                                                                                                                 | 0                             |
+| GIT_SPARSE_CHECKOUT            | `--sparse-checkout-file`         | the location of an optional [sparse-checkout](https://git-scm.com/docs/git-sparse-checkout#_sparse_checkout) file, same syntax as a .gitignore file.                                                                    | ""                             |
```
Comment (thockin) [inline README.md:97]: Sorry - code says "GIT_SYNC_SPARSE_CHECKOUT_FILE"  :(

Comment (SpencerMalone) [reply README.md]: You are 100% correct, I am clearly not awake enough to be making commits!

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve

The v4 dev branch (master) is somewhat divergent from v3, but this should port cleanly.  I'll take a quick look.

**Comment by thockin** [comment]
ehh, v4 needed a few tweaks.  PR coming

**Comment by thockin** [comment]
https://github.com/kubernetes/git-sync/releases/tag/v3.3.1 - waiting for the image to be promoted.

**Pull request merged**